# wf-mapping

Nexftlow workflow for basecalling and mapping of Nanopore fastq reads using dorado, minimap2, samtools and multiqc.

Requires either [Docker] or [Apptainer] installed.

Usage:

```sh
nextflow run chusj-pigu/wf-mapping [--pod5 INPUT] [--ref REF_GENOME] [OPTIONS]
```

To list available options and parameters for this workflow, run :
``` sh
nextflow run chusj-pigu/wf-mapping --help
```

Overview:

This workflow can be run locally or on Compute Canada. To use on Compute Canada, use the `-profile drac` option. For testing purposes, basecalling can be run on cpu only with `-profile no_gpu`. 

## 1. Basecalling
Basecalling is done with [Dorado] on duplex mode with modified base calling by default, using the Super accurate algorithm. The pod5 files are first split by channel to facilitate duplex calling.

## 2. Mapping
Mapping of the reads is done with [minimap2] using `-ax -map-ont` options for long reads generated by Nanopore sequencing. If `--m` was used, minimap2 will copy modified base information to output.

## 3. Sorting and indexing
Using [samtools], the .sam file is first converted to .bam using `samtools view`, sorted with `samtools sort` and indexed using `samtools index`. Statistics are reported with `samtools stats` and `multiqc`. 

## Outputs 
Generates a .sam file, .bam files for sorted and indexed reads. The fastp.json file for fastq statistics and .stats.txt file for mapping statistics are summarized in a [multiqc] report.




[Docker]: https://www.docker.com
[Apptainer]: https://apptainer.org
[Dorado]: https://github.com/nanoporetech/dorado
[minimap2]: https://lh3.github.io/minimap2/minimap2.html
[samtools]: http://www.htslib.org
[multiqc]: https://multiqc.info
[fastp]: https://github.com/OpenGene/fastp
