# wf-mapping

Nexftlow workflow for mapping of Nanopore fastq reads using minimap2, samtools and multiqc.

Requires either [Docker], [Singularity] or [Apptainer] installed.

Usage:

```sh
nextflow run chusj-pigu/wf-mapping [--reads INPUT] [--ref REF_GENOME] [OPTIONS]
```

To list available options and parameters for this workflow, run `nextflow run chusj-pigu/wf-mapping --help`

Overview:

## 1. FASTQ pre-processing
If the input is an unaligned uBAM file, it will be converted to FASTQ prior to mapping (`--bam` option) using `samtools fastq`. If the input is already in FASTQ format, this step will be skipped.
If `--m` was used, [samtools] will conserve all tags present in the uBAM file. 
General statistics on fastq files are obtained with [fastp].

## 2. Mapping
Mapping of the reads is done with [minimap2] using `-ax -map-ont` options for long reads generated by Nanopore sequencing. If `--m` was used, minimap2 will copy modified base information to output.

## 3. Sorting and indexing
Using [samtools], the .sam file is first converted to .bam using `samtools view`, sorted with `samtools sort` and indexed using `samtools index`. Statistics are reported with `samtools stats` and `multiqc`. 

## Outputs 
Generates a .sam file, .bam files for sorted and indexed reads, .stats text file for general statistics of the mapping and a [multiqc] report.




[Docker]: https://www.docker.com
[Singularity]: https://docs.sylabs.io/guides/3.5/user-guide/introduction.html
[Apptainer]: https://apptainer.org
[minimap2]: https://lh3.github.io/minimap2/minimap2.html
[samtools]: http://www.htslib.org
[multiqc]: https://multiqc.info
[fastp]: https://github.com/OpenGene/fastp
